pyscripts/build_docker_image.py

```diff
@@ -10,193 +10,237 @@
-    print(print_separator)
-    api_client = docker.APIClient(base_url="unix://var/run/docker.sock")
-    for image_tag in tags.image_tags:
-        print("Pushing {}:{}".format(tags.docker_full_image_name, image_tag))
-        for line in api_client.push(
-            tags.docker_full_image_name, image_tag, stream=True, auth_config=auth_config
-        ):
-            print_filtered_docker_line(line)
-
-
-def prefetch_cached_images(tags):
-    if docker_tags.running_on_azure_pipelines():
-        print(print_separator)
-        print(Fore.YELLOW + "PREFETCHING IMAGE")
-        print(print_separator)
-        tags.image_tag_to_use_for_cache = None
-        api_client = docker.APIClient(base_url="unix://var/run/docker.sock")
-        for image_tag in tags.image_tags:
+                dockerfile = "Dockerfile"
+
             print(
                 Fore.YELLOW
-                + "trying to prefetch {}:{}".format(
-                    tags.docker_full_image_name, image_tag
-                )
+                + "Building image for "
+                + tags.docker_image_name
+                + " using "
+                + dockerfile
             )
-            try:
-                for line in api_client.pull(
-                    tags.docker_full_image_name,
-                    image_tag,
-                    stream=True,
-                    auth_config=auth_config,
+            for line in api_client.build(
+                path=get_dockerfile_directory(tags),
+                tag=tags.docker_image_name,
+                buildargs=build_args,
+                cache_from=cache_from,
+                dockerfile=dockerfile,
+            ):
+                print_filtered_docker_line(line)
+
+        def tag_images(tags):
+            print(print_separator)
+            print("TAGGING IMAGE")
+            print(print_separator)
+
+            if sys.platform == 'win32':
+                base_url = "tcp://127.0.0.1:2375"
+            else:
+                base_url = "unix://var/run/docker.sock"
+            api_client = docker.APIClient(base_url=base_url)
+
+            print("Adding tags")
+            for image_tag in tags.image_tags:
+                print("Adding " + image_tag)
+                api_client.tag(tags.docker_image_name, tags.docker_full_image_name, image_tag.lower())
+
+
+        def push_images(tags):
+            print(print_separator)
+            print("PUSHING IMAGE")
+            print(print_separator)
+
+            if sys.platform == 'win32':
+                base_url = "tcp://127.0.0.1:2375"
+            else:
+                base_url = "unix://var/run/docker.sock"
+            api_client = docker.APIClient(base_url=base_url)
+
+            for image_tag in tags.image_tags:
+                print("Pushing {}:{}".format(tags.docker_full_image_name, image_tag))
+                for line in api_client.push(
+                    tags.docker_full_image_name, image_tag.lower(), stream=True, auth_config=auth_config
```

Comment:
>lower [](start = 59, length = 5)

Why did you need to add the "lower" call here?  I'm guessing that someone somewhere doesn't like mixed case, but who and why?  (if we're making bad tags, we should change docker_tags.py to not make bad tags).

Reply:
In the very first run, BEFORE image is in cache:
tag-image FAILS with error 500 due to the PascalCase used in get_commit_name() in docker_tags.py (line:63)
This is passed to image_tag api_client.tag(tags.docker_image_name, tags.docker_full_image_name, image_tag)
api_client.tag ONLY accepts tag in lower case.
Workaround was to add .lower() to image_tag in tag-image AND push-image in build_docker_image.py
Once the image is tagged with lower case, next run can continue using PascalCase because image is in cache and get-image is not case sensitive.
I can add the .lower() in docker_tags.py:get_commit_name(). This may be a better solution.

Reply:
please pull out the .lower changes. I don't feel comfortable with them.

---
In reply to: [295624635](https://github.com/Azure/iot-sdks-e2e-fx/pull/79#discussion_r295624635) [](ancestors = 295624635)

Reply:
Done.